Implement Vector{Size}<T>.AllBitsSet

Gnbrkm41 opened the pull request:
Resolves https://github.com/dotnet/runtime/issues/30659

Note: I strongly advise you to ignore the changes made in c9bd4a9 "Auto-generate tests": They are automatically generated from the template and adds lots of LoCs.

Gnbrkm41 committed:
Implement managed fallback of AllBitsSet

Gnbrkm41 committed:
Add JIT test templates

Gnbrkm41 committed:
Auto-generate tests

Gnbrkm41 committed:
Update test projects to use wildcards

Gnbrkm41 committed:
Intrinsify AllBitsSet

Gnbrkm41 commented:
cc @tannergooding

Gnbrkm41 commented:
I believe we can use this in a couple places:
https://github.com/dotnet/runtime/blob/963b1587b48ec2fa09cb5bcd0a53460762a459e4/src/libraries/System.Collections/src/System/Collections/BitArray.cs#L542-L556

Gnbrkm41 commented:
All of the failures seem to be either of those:
```
Assertion failed '!"Jump into middle of try region"' in 'JIT.HardwareIntrinsics.General.Program:Vector64AllBitsSet()' during 'Optimize layout' (IL size 50) File: F:\workspace\_work\1\s\src\coreclr\src\jit\flowgraph.cpp Line: 20738\r\n Image: C:\h\w\AC45098D\p\CoreRun.exe\r\n\r\n\r\n
```
https://dev.azure.com/dnceng/public/_build/results?buildId=568937&view=ms.vss-test-web.build-test-results-tab&runId=17857130&resultId=102291&paneView=debug
```
Assert failure(PID 1220 [0x000004c4], Thread: 5788 [0x169c]): Assertion failed 'fixedArity == 0' in 'JIT.HardwareIntrinsics.General.Program:AllBitsSetByte()' during 'Do value numbering' (IL size 38)\r\n\r\n File: F:\workspace\_work\1\s\src\coreclr\src\jit\valuenum.cpp Line: 8588
```
https://dev.azure.com/dnceng/public/_build/results?buildId=568937&view=ms.vss-test-web.build-test-results-tab&runId=17857130&resultId=102416&paneView=debug

gfoidl commented:
> Vector256<int> ones = Vector256.Create(-1);

Would it be possible for the JIT to pattern match such cases and emit the `cmpps`?
If not or additional it would be great to have an analyzer and codefix that changes this to `Vector256<int>.AllBitsSet`.

Gnbrkm41 commented:
Weird that I do not reproduce the test failures on my local machine :^(

Gnbrkm41 commented:
So, it appears that this is happening because methods that can have different instructions depending on types need an extra `VNF_SimdType` arg:
https://github.com/dotnet/runtime/blob/24c4cb193fdf6091fe07695d81b182674e2c1629/src/coreclr/src/jit/hwintrinsic.cpp#L284

What does this really mean?

tannergooding commented:
> So, it appears that this is happening because methods that can have different instructions depending on types need an extra VNF_SimdType arg

CC. @briansull who added the initial VN support for HWIntrinsics in https://github.com/dotnet/runtime/pull/31834

tannergooding commented:
Also CC. @echesakovMSFT and @CarolEidt

briansull commented:
> need an extra VNF_SimdType arg:

This prevents us from CSE-ing two different SIMD e=operations that would be implemented using different instructions.   I do not think that it is the reason for the assert 

`Assertion failed 'fixedArity == 0'`

That assert is likely caused by a -1 or 0 entry in the SIMD instruction table.

tannergooding commented on src/coreclr/src/jit/hwintrinsiccodegenxarch.cpp:
```diff
@@ -1239,6 +1239,24 @@ void CodeGen::genBaseIntrinsic(GenTreeHWIntrinsic* node)
+            else
+            {
+                assert(varTypeIsIntegral(baseType) || !compiler->compSupports(InstructionSet_AVX));
+                emit->emitIns_SIMD_R_R_R(ins, attr, targetReg, targetReg, targetReg);
```
What is used for the comparison for `float/double` when AVX isn't supported? Based on the instruction table, this is `cmpps/cmppd` still; but I don't think that is correct?

Gnbrkm41 replied:
Ohhhh, good catch.

I wonder how we'd change it here, though? hard-code the instruction here?

tannergooding replied:
We would need to special-case one of the paths.

That's also possibly an interesting case for CSE... If the table gives a particular set of instructions but codegen can optionally special-case something further or treat it slightly differently, how should that be handled @briansull ?

briansull replied:
The ValueNumber & CSE phase only uses the table to determine if the result type needs to be an input when generating the ValueNumber for the node.

If for two trees that have the same operation (i.e. GT_MUL or the same HW intrinsic) and  all of its operands have the identical value numbers, then normally we would give the same value number.  

For GT_CAST we incorporate the castto type as an extra operand to the value number.

I noticed that we also needed to do this for some SIMD and HW instrinsic nodes.
I determined that safest and easiest way to to do this was to examine the table of instructions and always incorporate the result type when there were two or more different valid instructions listed for a SIMD or HW instrinsic node.

This process is used for x86 and x64, I believe that we need to be more conservative on ARM64, so we always include an extra result type operand.

As as long as the table has two or more different instructions we will be good.
It would be bad to list all the same instructions or an illegal instruction and then use hand code logic to decide on the instruction.   If you want to record that an entry relies upon hand coded logic I would recommend using a brkpt or nop instruction as a marker for this behavior in the table.  We can add a check for this and assume that different instructions could be generated.  I don't think that it should matter if AVX is supported or not when deciding if we need an extra result type operand.

tannergooding replied:
> I don't think that it should matter if AVX is supported or not when deciding if we need an extra result type operand.

This would be a case of, when not using `VEX` encoding (SSE-SSE4.2) all types would use the same instruction. But when using the `VEX` encoding (AVX+), `float`/`double` would use different instructions that are more efficient.

Gnbrkm41 replied:
Left a comment that float/double may use different instructions depending on the encoding available. Also hardcoded the instruction inside the integer / non-VEX path.

tannergooding commented:
> I will investigate the assert and provide guidance
> `Assertion failed 'fixedArity == 0'`

I did some initial debugging and found that `Vector128_Zero` doesn't hit this issue as it has a single instruction (`xorps`) used by all types (and that looks to be the case for all other `0` arg intrinsics right now).

`Vector128_AllBitSet` has three different instructions and so it gets the additional `VNF_SimdType` arg and the arity becomes `1`, failing the assert.

briansull commented:
The fix is:

line 8569 in valuenum.cpp in function:
`void Compiler::fgValueNumberHWIntrinsic(GenTree* tree)`

```
    int      lookupNumArgs    = HWIntrinsicInfo::lookupNumArgs(hwIntrinsicNode->gtHWIntrinsicId);
    bool     encodeResultType = vnEncodesResultTypeForHWIntrinsic(hwIntrinsicNode->gtHWIntrinsicId);
    VNFunc   func             = GetVNFuncForNode(tree);
    unsigned fixedArity       = vnStore->VNFuncArity(func);

    ValueNumPair excSetPair = ValueNumStore::VNPForEmptyExcSet();
    ValueNumPair normalPair;
    ValueNumPair resvnp     = ValueNumPair();

    if (encodeResultType)
    {
        ValueNum vnSize = vnStore->VNForIntCon(hwIntrinsicNode->gtSIMDSize);
        ValueNum vnBaseType = vnStore->VNForIntCon(INT32(hwIntrinsicNode->gtSIMDBaseType));
        ValueNum simdTypeVN = vnStore->VNForFunc(TYP_REF, VNF_SimdType, vnSize, vnBaseType);
        resvnp.SetBoth(simdTypeVN);

#ifdef DEBUG
        if (verbose)
        {
            printf("    simdTypeVN is ");
            vnPrint(simdTypeVN, 1);
            printf("\n");
        }
#endif
    }

    // There are some HWINTRINSICS operations that have zero args, i.e.  NI_Vector128_Zero
    if (tree->AsOp()->gtOp1 == nullptr)
    {
        if (encodeResultType)
        {
            // There are zero arg HWINTRINSICS operations that encode the result type, i.e.  Vector128_AllBitSet 
            normalPair = vnStore->VNPairForFunc(tree->TypeGet(), func, resvnp);
            assert(vnStore->VNFuncArity(func) == 0);
        }
        else
        {
            normalPair = vnStore->VNPairForFunc(tree->TypeGet(), func);
            assert(vnStore->VNFuncArity(func) == 1);
        }

    }
    else if (tree->AsOp()->gtOp1->OperIs(GT_LIST) || (lookupNumArgs == -1))
```

Gnbrkm41 commented:
```
*************** Starting PHASE Merge throw blocks

*************** In fgTailMergeThrows

*** Does not return call
               [000327] --CXG+------              *  CALL      void   System.ThrowHelper.ThrowNotSupportedException
               [000326] -----+------ arg0 in rcx  \--*  CNS_INT   int    63
    in BB04 is unique, marking it as canonical

*** Does not return call
               [000151] --CXG+------              *  CALL      void   System.ThrowHelper.ThrowNotSupportedException
               [000150] -----+------ arg0 in rcx  \--*  CNS_INT   int    63
    in BB02 can be dup'd to canonical BB04

*** found 1 merge candidates, rewriting flow

New Basic Block BB10 [0053] created.
*** BB01 now falling through to empty BB10 and then to BB04

*************** After fgTailMergeThrows(1 updates)

-----------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight    lp [IL range]     [jump]      [EH region]         [flags]
-----------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1        [000..002)                                     i label target
BB10 [0053]  1       BB01                  1        [???..???)-> BB04 (always)                     internal
BB02 [0001]  0  0                          0        [002..00A)        (throw ) T0      try {       keep i try rare label gcsafe
BB03 [0019]  0  0                          1        [002..003)                 T0                  keep i label gcsafe
BB04 [0033]  2  0    BB03,BB10             0        [002..003)        (throw ) T0                  keep i rare label target gcsafe
BB05 [0046]  0  0                          1        [???..???)-> BB07 (always) T0      }           keep i internal label
BB07 [0003]  2       BB05,BB06             1        [00F..012)-> BB09 ( cond )                     i label target
BB08 [0004]  1       BB07                  0        [012..031)        (throw )                     i rare gcsafe newobj
BB09 [0005]  1       BB07                  1        [031..032)        (return)                     i label target
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ funclets follow
BB06 [0002]  1     0                       1        [00A..00F)-> BB07 ( cret )    H0 F catch { }   keep i label target flet
-----------------------------------------------------------------------------------------------------------------------------------------
*************** In fgDebugCheckBBlist
Jump into the middle of try region: BB10 branches to BB04

Assert failure(PID 11372 [0x00002c6c], Thread: 3996 [0x0f9c]): Assertion failed '!"Jump into middle of try region"' in 'JIT.HardwareIntrinsics.General.Program:Vector64AllBitsSet()' during 'Merge throw blocks' (IL size 50)

    File: C:\Users\gotos\source\repos\runtime\src\coreclr\src\jit\flowgraph.cpp Line: 20738
    Image: c:\users\gotos\source\repos\runtime\artifacts\tests\coreclr\windows_nt.x64.checked\tests\core_root\corerun.exe
```
[Full JitDump](https://github.com/dotnet/runtime/files/4376775/log.txt)

An interesting failure...

AndyAyersMS commented:
@Gnbrkm41 that's a new phase I added recently, probably a missing safety check. Let me take a look.

AndyAyersMS commented:
`fgTailMergeThrowsFallThroughHelper` needs to ensure that the new BB is in the same EH region as the the "nonCanonicalBlock". It is not doing this properly if the nonCanonicalBlock is a try entry.

Might be simplest for now to disable throw helper merging in this case. Will keep looking.

AndyAyersMS commented:
@Gnbrkm41 see if this patch fixes your failures. Haven't validated it yet -- trying to create a simple local repro, but no luck so far.
```diff
index 8ba6d2f8cc7..846677f482f 100644
--- a/src/coreclr/src/jit/flowgraph.cpp
+++ b/src/coreclr/src/jit/flowgraph.cpp
@@ -25869,6 +25869,14 @@ void Compiler::fgTailMergeThrows()
     // and there is less jumbled flow to sort out later.
     for (BasicBlock* block = fgLastBB; block != nullptr; block = block->bbPrev)
     {
+        // Workaround: don't consider try entry blocks as candidates
+        // for merging; if the canonical throw is later in the same try,
+        // we'll create invalid flow.
+        if ((block->bbFlags & BBF_TRY_BEG) != 0)
+        {
+            continue;
+        }
+
         // For throw helpers the block should have exactly one statement....
         // (this isn't guaranteed, but seems likely)
         Statement* stmt = block->firstStmt();
```

AndyAyersMS commented:
Ok, I can repro now. I'll put up a fix.

Gnbrkm41 committed:
Fix value numbering

Co-Authored-By: Brian Sullivan <briansul@microsoft.com>

Gnbrkm41 committed:
Apply formatting

Gnbrkm41 committed:
Special-case the integer path

BruceForstall commented:
Does this need an arm64 implementation?

tannergooding commented:
> Does this need an arm64 implementation?

It will need one for `Vector64<T>` and `Vector128<T>`. @TamarChristinaArm could you advise as to which instructions should be used to efficiently create a `vector` where all bits are set?
That is, should we do a set and duplicate or maybe a `compare tgt, tgt, tgt` like we do on x86, or maybe something else?

TamarChristinaArm commented:
> > Does this need an arm64 implementation?
> 
> It will need one for `Vector64<T>` and `Vector128<T>`. @TamarChristinaArm could you advise as to which instructions should be used to efficiently create a `vector` where all bits are set?
> That is, should we do a set and duplicate or maybe a `compare tgt, tgt, tgt` like we do on x86, or maybe something else?

@tannergooding  you have a couple of options:

```
mvni v0.4s, #0
movi v1.16b, #0xFF
```

will both give you a vector with all bits set.

`mvni` gives you some slight flexibility in if you want to set only the bottom bits you can do `mvni v0.2s, #0`.

Gnbrkm41 commented:
Note that I got slightly busy recently. I hope I have some time to work on intrinsifying ARM64, but I am not sure how quick can happen; Do you think it'll be okay if I can attempt the ARM part later (hopefully soon, next weekend?) and open a follow up PR?

Gnbrkm41 commented:
Rebased and immediately there's a merge conflict.... I'll look into it.

briansull commented:
I just checked in the fixes that I shared with you for ValueNumbering

tannergooding commented:
@davidwrighton, do we have a reference doc for which APIs to use instead of `compSupports` now that it has been changed?
(I don't recall if that was in the list of docs that were added)

davidwrighton commented:
Yes, there is a BotR chapter that I wrote. See the end of https://github.com/dotnet/runtime/blob/master/docs/design/coreclr/botr/vectors-and-intrinsics.md 

You'll want `compExactlyDependsOn` or `compOpportunisticallyDependsOn`. For this code I think you'll want `compExactlyDependsOn`.

Gnbrkm41 committed:
Revert "Fix value numbering"

This reverts commit a708e533368cb6b8f71aa0ada3d931611dea1722.

Gnbrkm41 committed:
Use new ISA query methods

Gnbrkm41 committed:
Use compIsaSupportedDebugOnly in asserts

* compExactlyDependsOn(isa) ... Should never be used in an assert.

tannergooding commented:
> You'll want compExactlyDependsOn or compOpportunisticallyDependsOn. For this code I think you'll want compExactlyDependsOn.

@davidwrighton, why `Exactly` here? Based on https://github.com/dotnet/runtime/pull/34204#discussion_r407590582, this should be `Opportunistically` since the two paths have equivalent semantics, they just use different code to generate a vector with all bits set, depending on what ISAs are available. Is that correct?

davidwrighton commented:
@tannergooding Ah, you're right. I misread the change when I looked at it earlier and thought it would affect the behavior of Vector256 on SSE only machines to see AVX vs not, but upon reflection, Vector256 cases won't reach this code on SSE only compile, so `Opportunistically` is useable.

Gnbrkm41 committed:
Use opportunistically depends on instead

Gnbrkm41 commented:
*I think* this is ready for x86. I can look into ARM ones later (hopefully).

@TamarChristinaArm, Out of curiosity, is there a way you would recommend for creating a vector with all bits set, specifically for floating point types? I heard that in x86 platform creating an integer vector then re-interpreting it as a floating point vector can cause pipeline switch and (a little bit) of performance degradation. I wonder if this is the case in ARM processors as well.

TamarChristinaArm commented:
> @TamarChristinaArm, Out of curiosity, is there a way you would recommend for creating a vector with all bits set, specifically for floating point types? I heard that in x86 platform creating an integer vector then re-interpreting it as a floating point vector can cause pipeline switch causing (a little bit of) performance degradation. I wonder if this is the case in ARM processors as well.

@Gnbrkm41 The same sequences as above work with no additional overhead. In Armv8 There's only one SIMD/FP register file and it's the instructions that interpret the bits in the vector. So there's no overhead to interpret an integer vector as a floating point vector. Scalars can have depending on where the scalars live. an integer scalar can live in both register files so if you do a conversion you may have to move it.

Gnbrkm41 commented:
If all goes well, I think I should be able to update this PR to use ARM instructions soon.

Gnbrkm41 committed:
Intrinsify Zero and AllBitsSet for ARM

Gnbrkm41 commented:
Rebased & Updated to use ARM instructions for get_Zero and get_AllBitsSet: 
![unknown2](https://user-images.githubusercontent.com/42944058/79671100-afd8fc00-8202-11ea-8ff4-4dbc4aefcb5d.png)
![unknown](https://user-images.githubusercontent.com/42944058/79671101-b10a2900-8202-11ea-88f9-496e70b7aa2b.png)

(The latest implementation always uses 2S/4S variant depending on the vector size)

Gnbrkm41 commented on src/coreclr/src/jit/hwintrinsiclistarm64.h:
```diff
@@ -42,6 +44,8 @@ HARDWARE_INTRINSIC(Vector128,       AsUInt16,                                  -
 HARDWARE_INTRINSIC(Vector128,       AsUInt32,                                  -1,              16,           1,     {INS_invalid,           INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid},           HW_Category_Helper,                 HW_Flag_SpecialImport|HW_Flag_NoCodeGe ...
 HARDWARE_INTRINSIC(Vector128,       AsUInt64,                                  -1,              16,           1,     {INS_invalid,           INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid},           HW_Category_Helper,                 HW_Flag_SpecialImport|HW_Flag_NoCodeGe ...
 HARDWARE_INTRINSIC(Vector128,       get_Count,                                 -1,              16,           0,     {INS_invalid,           INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid},           HW_Category_Helper,                 HW_Flag_SpecialImport|HW_Flag_NoCodeGe ...
+HARDWARE_INTRINSIC(Vector128,       get_Zero,                                   0,              16,           0,     {INS_movi,              INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi},              HW_Category_Helper,                 HW_Flag_SpecialImport)
```
Could someone see if I am not missing any additional flags that apply to this particular one? (I likely do)

echesakovMSFT commented on src/coreclr/src/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -319,6 +326,15 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+            // mvni won't take 8B/16B, so hard-code the size and reinterpret the bits. The behaviour is the same.
+            case NI_Vector64_get_Zero:
+            case NI_Vector64_get_AllBitsSet:
+                GetEmitter()->emitIns_R_I(ins, emitSize, targetReg, ival, INS_OPTS_2S);
```
Can you please hard-code `ival` value here and remove call to `HWIntrinsicInfo::lookupIval` - I am removing `HWIntrinsicInfo::lookupIval` and the `ival` column from hwintrinsiclistarm64.h on Arm64 in #35030

echesakovMSFT commented on src/coreclr/src/jit/hwintrinsicarm64.cpp:
```diff
@@ -337,6 +337,16 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
+        case NI_Vector128_get_Zero:
+        case NI_Vector64_get_AllBitsSet:
+        case NI_Vector128_get_AllBitsSet:
+        {
```
You probably don't need this - if you change a category to SimpleSIMD the importation should be automatically handled as I said below

echesakovMSFT commented on src/coreclr/src/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -319,6 +326,15 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+            case NI_Vector64_get_Zero:
+            case NI_Vector64_get_AllBitsSet:
+                GetEmitter()->emitIns_R_I(ins, emitSize, targetReg, ival, INS_OPTS_2S);
+                break;
```
Can you please put an empty line after `brake` here

echesakovMSFT commented on src/coreclr/src/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -319,6 +326,15 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
+            case NI_Vector128_get_Zero:
+            case NI_Vector128_get_AllBitsSet:
+                GetEmitter()->emitIns_R_I(ins, emitSize, targetReg, ival, INS_OPTS_4S);
+                break;
```
Can you please put an empty line after brake here as well

echesakovMSFT commented on src/coreclr/src/jit/hwintrinsiclistarm64.h:
```diff
@@ -24,6 +24,8 @@ HARDWARE_INTRINSIC(Vector64,        AsSingle,                                  -
 HARDWARE_INTRINSIC(Vector64,        AsUInt16,                                  -1,               8,           1,     {INS_invalid,           INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid},           HW_Category_Helper,                 HW_Flag_SpecialImport|HW_Flag_NoCodeGe ...
 HARDWARE_INTRINSIC(Vector64,        AsUInt32,                                  -1,               8,           1,     {INS_invalid,           INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid},           HW_Category_Helper,                 HW_Flag_SpecialImport|HW_Flag_NoCodeGe ...
 HARDWARE_INTRINSIC(Vector64,        get_Count,                                 -1,               8,           0,     {INS_invalid,           INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid},           HW_Category_Helper,                 HW_Flag_SpecialImport|HW_Flag_NoCodeGe ...
+HARDWARE_INTRINSIC(Vector64,        get_Zero,                                   0,               8,           0,     {INS_movi,              INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi,           INS_movi},              HW_Category_Helper,                 HW_Flag_SpecialImport)
```
I believe if you change category to `HW_Category_SimpleSIMD` you wouldn't need the custom importation logic - but you would need `HW_Flag_SpecialCodeGen` (the only flag that is needed here)

Gnbrkm41 replied:
In the latest commit, I've changed the category to SimpleSIMD and removed the custom importation logic. However, it looks like it is still hitting the `impSpecialCodeGen` method and now because I removed the importation logic the call is not imported as a `GT_HWINTRINSIC` node. Jit dump here:
```
Trees after linear scan register allocator (LSRA)

-----------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight    lp [IL range]     [jump]      [EH region]         [flags]
-----------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1        [000..01D)        (return)                     i label target gcsafe LIR
-----------------------------------------------------------------------------------------------------------------------------------------

------------ BB01 [000..01D) (return), preds={} succs={}
N002 (???,???) [000021] ------------                 IL_OFFSET void   IL offset: 0x0 REG NA
N004 (  1,  1) [000013] ------------        t13 =    CNS_INT(h) long   0x4001BB8 [ICON_STR_HDL] REG x0
                                                  ┌──▌  t13    long
N006 (  4,  3) [000014] n---G-------        t14 = ▌  IND       ref    REG x0
                                                  ┌──▌  t14    ref
N008 (???,???) [000025] ----G-------        t25 = ▌  PUTARG_REG ref    REG x0
                                                  ┌──▌  t25    ref    arg0 in x0
N010 ( 18,  6) [000001] --CXG-------              ▌  CALL      void   TestLibrary.TestFramework.BeginScenario REG NA
N012 (???,???) [000022] ------------                 IL_OFFSET void   IL offset: 0xa REG NA
N014 ( 14,  2) [000002] --CXG-------         t2 =    CALL      simd16 System.Runtime.Intrinsics.Vector128`1[UInt16][System.UInt16].get_AllBitsSet REG d0
                                                  ┌──▌  t2     simd16
N016 ( 18,  5) [000005] DA-XG-------              ▌  STORE_LCL_VAR simd16<System.Runtime.Intrinsics.Vector128`1[UInt16]> V01 loc0          NA REG NA
N018 (???,???) [000023] ------------                 IL_OFFSET void   IL offset: 0x10 REG NA
N020 (  1,  1) [000016] ------------        t16 =    CNS_INT(h) long   0x4001BB8 [ICON_STR_HDL] REG x1
                                                  ┌──▌  t16    long
N022 (  4,  3) [000017] n---G-------        t17 = ▌  IND       ref    REG x1
                                                  ┌──▌  t17    ref
N024 (???,???) [000026] ----G-------        t26 = ▌  PUTARG_REG ref    REG x1
N026 (  3,  2) [000006] ------------         t6 =    LCL_VAR   ref    V00 this          x0 REG x0
                                                  ┌──▌  t6     ref
N028 (???,???) [000027] ------------        t27 = ▌  PUTARG_REG ref    REG x0
N030 (  3,  2) [000007] ------------         t7 =    LCL_VAR   simd16<System.Runtime.Intrinsics.Vector128`1[UInt16]> V01 loc0          d0 REG d0
                                                  ┌──▌  t7     simd16
N032 (???,???) [000028] ------------        t28 = ▌  PUTARG_REG simd16 REG d0
                                                  ┌──▌  t26    ref    arg2 in x1
                                                  ├──▌  t27    ref    this in x0
                                                  ├──▌  t28    simd16 arg1 in d0
N034 ( 24, 13) [000009] --CXG-------              ▌  CALL      void   JIT.HardwareIntrinsics.General.VectorAllBitsSet__AllBitsSetUInt16.ValidateResult REG NA
N036 (???,???) [000024] ------------                 IL_OFFSET void   IL offset: 0x1c REG NA
N038 (  0,  0) [000012] ------------                 RETURN    void   REG NA

-------------------------------------------------------------------------------------------------------------------
```

Gnbrkm41 replied:
(@echesakovMSFT)

echesakovMSFT replied:
@Gnbrkm41 You changed the category but kept SpecialImport flag. What I have suggested is to change the flags column so it has only SpecialCodeGen instead

> if you change category to HW_Category_SimpleSIMD you wouldn't need the custom importation logic - but you would need HW_Flag_SpecialCodeGen

I guess the way I worded this was confusing...

@tannergooding I wonder if in case when SpecialImport flag is set but there is no custom importation logic we should assert instead of expanding an intrinsic as a call?

Gnbrkm41 replied:
ah, right, sorry, I must have missed that part 😅

tannergooding replied:
The category and helpers are different and have slightly different meanings iirc.
We could probably simplify it a bit, but its likely better to do in a separate PR.

Gnbrkm41 replied:
RE: custom importation logic - I've managed to fix the assertion failures for the expected types, but now having problem with invalid base types, as the API allows any type for the base type (e.g. `Vector64<bool>.Zero`), and the importing logic for table driven intrinsics doesn't handle arbitrary base types.

Do we modify the `impHWIntrinsic` method to handle that, or just write custom import logic? those two APIs (Zero, AllBitsSet) are kind of a special case so it feels right to just make it `SpecialImport`...

Gnbrkm41 replied:
Actually, I am not sure what is the problem here. Vector64<bool>.Zero works just fine but the 128 one fails during import.

tannergooding replied:
Taking a look.

tannergooding replied:
`Vector64<T>` isn't failing because none of the methods are marked `[Intrinsic]` as they are on `Vector128<T>`.

`Vector128<T>` is asserting in `HWIntrinsicInfo::lookupIns` because the `type` isn't one of the 10 supported. `impHWIntrinsic` needs something like the following:
```c++
if (!varTypeIsArithmetic(baseType))
{
    return nullptr;
}
```

It should probably be inserted here, right after we finish determining the baseType: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/hwintrinsic.cpp#L645
* Note: I didn't do any extensive analysis to make sure we aren't missing something for the non-table driven intrinsics

Gnbrkm41 replied:
> `Vector64<T>` isn't failing because none of the methods are marked [Intrinsic] as they are on `Vector128<T>`. 

Aaaaaaaaah. that explains why I was not hitting any of the breakpoints under `impIntrinsic` when importing Vector64 methods. I never checked the disasm for Vector64 methods and they apparently are not properly intrinsified also... 😅

Gnbrkm41 replied:
> `impHWIntrinsic` needs something like the following:

Do we want to do this, though? for me it seems to suggest that we do need the special importing logics for it since none of the types expect arbitrary base types to be specified...

echesakovMSFT replied:
@Gnbrkm41 I personally would be fine with either approach.

I would add the check that Tanner suggested anyway - this will ensure that if the base type does not correspond to a valid element type of a SIMD vector then treat the method being imported as a simple call not as an intrinsic - this what "return nullptr" technically means.

Gnbrkm41 replied:
The check is technically already there, in this part:
```cpp
if (!isScalar && ((HWIntrinsicInfo::lookupIns(intrinsic, baseType) == INS_invalid) ||
                    ((simdSize != 8) && (simdSize != 16) && (simdSize != 32))))
{
    assert(!"Unexpected HW Intrinsic");
    return nullptr;
}
```
`HWIntrinsicInfo::lookupIns` asserts that `baseType` is one of the supported types, and additionally returns `INS_invalid` if `baseType` is invalid. So we already do have the check (we won't hard crash in release builds during JIT because of it), it's just that we assume that none of the table driven importations can have invalid base type and what we have now causes assert failure in those places.

tannergooding replied:
The assert is there to avoid errors later down in the stack. I think it is fine to duplicate it so we can otherwise maintain the assertion.

Gnbrkm41 replied:
After doing this, the CLR started hitting an assertion in `CLRException::GetThrowable` ("Recursion in CLRException::GetThrowable"); perhaps the changes we introduced here are returning nullptr too early on invalid baseTypes? I think we need to investigate a little bit more if we want to make those table-driven. For now, marking those as `SpecialImport` and re-introducing the custom importation logics.
![image](https://user-images.githubusercontent.com/42944058/80291603-ecb86c00-8789-11ea-8205-04273a69a8da.png)

kunalspathak replied:
This is interesting. I have started noticing this recently.

Gnbrkm41 replied:
The change that caused the assertion failure was reverted, so I doubt this PR contains the cause for that exception / is relevant?

I'd be interested to know what exactly is causing that though.

kunalspathak replied:
Sure, not blaming this PR :), but there is something that we both did and causing this. This is my callstack.

![image](https://user-images.githubusercontent.com/12488060/81242814-f4b2b000-8fc2-11ea-8dba-fcc811a046fa.png)

echesakovMSFT commented on src/coreclr/src/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -140,7 +144,9 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
             op1Reg = intrin.op1->GetRegNum();
             break;
-
+        case 0:
```
Please put empty lines before and after your code

echesakov reviewed (commented):
I reviewed arm64 part

Gnbrkm41 committed:
Address PR feedback

* Hard-code ivals
* Remove custom importation logic (and replace with HW_Category_SimpleSIMD)
* Insert newlines

Gnbrkm41 committed:
Mark Zero and AllBitsSet as SpecialCodeGen

Gnbrkm41 committed:
Allow null op1s where needed

Gnbrkm41 committed:
Fix failures for unsupported types, mark methods as intrinsic

CarolEidt commented on src/coreclr/src/jit/hwintrinsic.cpp:
```diff
@@ -643,6 +643,11 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
     baseType          = getBaseTypeFromArgIfNeeded(intrinsic, clsHnd, sig, baseType);
     unsigned simdSize = HWIntrinsicInfo::lookupSimdSize(this, intrinsic, sig);
 
+    if (!varTypeIsArithmetic(baseType))
```
@tannergooding - I see that you recommended this. Can you explain why this wasn't needed previously?

tannergooding replied:
The majority of intrinsics aren't generic and we were treating the couple that were (`Vector128<T>.Count` and `Vector128<T>.Zero`) as `SpecialImport`.

This allows them to be table driven instead as it handles the case where a user tries `Vector128<bool>.Zero`, for example.

CarolEidt replied:
Great - thanks for clarifying

Gnbrkm41 replied:
This change is no longer existent, as we marked this as `SpecialImport` again.

CarolEidt commented on src/coreclr/src/jit/hwintrinsiccodegenarm64.cpp:
```diff
@@ -449,6 +452,17 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                 }
                 break;
 
+            // mvni won't take 8B/16B, so hard-code the size and reinterpret the bits. The behaviour is the same.
```
In what way is the size being reinterpreted here?

Gnbrkm41 replied:
As in, we'll always use instructions for 32-bit sized elements [because MVNI doesn't allow 8/16 1-byte elements.](https://developer.arm.com/docs/ddi0596/e/simd-and-floating-point-instructions-alphabetic-order/mvni-move-inverted-immediate-vector). Behaviour-wise they are the same (because we invert no bits or all the bits).

CarolEidt replied:
Ah, OK - might be better worded as "mvni doesn't support the range of element types, so hard code the 'opts' value." - I was confused into thinking that it was the `emitSize` that it was talking about.

CarolEidt commented on src/coreclr/src/jit/hwintrinsiclistxarch.h:
```diff
@@ -44,6 +44,8 @@ HARDWARE_INTRINSIC(Vector128,       AsVector3,
 HARDWARE_INTRINSIC(Vector128,       AsVector4,                                  16,             1,      {INS_invalid,           INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid},           HW_Category_Helper,                 HW_Flag_SpecialImpo ...
 HARDWARE_INTRINSIC(Vector128,       AsVector128,                                16,             1,      {INS_invalid,           INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid},           HW_Category_Helper,                 HW_Flag_SpecialImpo ...
 HARDWARE_INTRINSIC(Vector128,       CreateScalarUnsafe,                         16,             1,      {INS_mov_i2xmm,         INS_mov_i2xmm,          INS_mov_i2xmm,          INS_mov_i2xmm,          INS_mov_i2xmm,          INS_mov_i2xmm,          INS_mov_i2xmm,          INS_mov_i2xmm,          INS_movss,              INS_movsdsse2},         HW_Category_SIMDScalar,             HW_Flag_SpecialImpo ...
+// The instruction generated for float/double depends on encodings available
```
It might be clearer here to say "depends on which which ISAs are supported".

CarolEidt commented on src/coreclr/src/jit/hwintrinsiclistxarch.h:
```diff
@@ -70,6 +72,8 @@ HARDWARE_INTRINSIC(Vector256,       AsUInt32,
 HARDWARE_INTRINSIC(Vector256,       AsUInt64,                                   32,             1,      {INS_invalid,           INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid},           HW_Category_Helper,                 HW_Flag_SpecialImpo ...
 HARDWARE_INTRINSIC(Vector256,       AsVector,                                   32,             1,      {INS_invalid,           INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid},           HW_Category_Helper,                 HW_Flag_SpecialImpo ...
 HARDWARE_INTRINSIC(Vector256,       AsVector256,                                32,             1,      {INS_invalid,           INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid,            INS_invalid},           HW_Category_Helper,                 HW_Flag_SpecialImpo ...
+// The instruction generated for float/double depends on encodings available
```
ditto here.

CarolEidt reviewed (commented):
Minor suggestions and questions.

Gnbrkm41 committed:
Fix compilation errors, edit comments per PR feedback

Gnbrkm41 committed:
Use custom importation logics for AllBitsSet/Zero on ARM

Gnbrkm41 committed:
Apply formatting

Gnbrkm41 committed:
Address PR feedback

Gnbrkm41 commented:
Just rebased, resolved conflicts and addressed feedbacks. I've locally checked that the changes for both ARM and xarch do in fact generate appropriate instructions and all the tests pass; Could I get a final review on this? Thanks!

tannergooding commented on src/coreclr/src/jit/hwintrinsicarm64.cpp:
```diff
@@ -281,7 +281,9 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
         if (!varTypeIsArithmetic(baseType))
         {
-            assert((intrinsic == NI_Vector64_AsByte) || (intrinsic == NI_Vector128_As));
+            assert((intrinsic == NI_Vector64_AsByte) || (intrinsic == NI_Vector128_As) ||
```
@echesakovMSFT, not related to this PR, but I think the first check in this is wrong. It should be `intrinsic == NI_Vector64_As`, right?

echesakovMSFT replied:
It looks so, however, I don't see Vector64.As intrinsic in hwintrinsiclistarm64.h, I will update #33308 to include this.

tannergooding reviewed (approved):
LGTM.

echesakov reviewed (approved):
LGTM

echesakovMSFT commented:
I did four attempts to re-run testing for this PR - **runtime (Installer Build and Test coreclr FreeBSD_x64 Debug)** keeps failing with 
```
/root/runtime/.dotnet/sdk/5.0.100-preview.4.20202.8/NuGet.RestoreEx.targets(10,5): error : No space left on device [/root/runtime/tools-local/tasks/installer.tasks/installer.tasks.csproj]
```
Everything else is green, even though is reported as "non finished" - https://dev.azure.com/dnceng/public/_build/results?buildId=621331&view=logs&j=41e34ca2-d347-5e65-d632-b45724e78141. Merging so it would not conflict with #35594.

Thanks @Gnbrkm41 for contribution!
